app/src/org/commcare/models/database/user/UserDbUpgradeUtils.java

```diff
@@ -198,4 +196,54 @@ protected static void migrateFormRecordsToV3(Context c, SQLiteDatabase db) {
             newStorage.write(newRecord);
         }
     }
+
+    protected static void migrateV4FormRecords(Context c, SQLiteDatabase db) {
+        SqlStorage<FormRecordV4> oldStorage = getFormRecordStorage(c, db, FormRecordV4.class);
+
+        // Transform old records to new add to newRecords
+        Vector<FormRecord> newRecords = new Vector<>();
+        for (FormRecordV4 oldRecord : oldStorage) {
+            FormRecord newRecord = new FormRecord(
+                    oldRecord.getStatus(),
+                    oldRecord.getFormNamespace(),
+                    oldRecord.getAesKey(),
+                    oldRecord.getInstanceID(),
+                    oldRecord.lastModified(),
+                    oldRecord.getAppId());
+
+            // Merge Record's instance
+            String instanceUri = oldRecord.getInstanceURIString();
+            Cursor cursor = c.getContentResolver().query(Uri.parse(instanceUri), null, null, null, null);
+            if (cursor != null && cursor.getCount() > 0) {
+                cursor.moveToFirst();
+                newRecord.setDisplayName(cursor.getString(cursor.getColumnIndex(InstanceProviderAPI.InstanceColumns.DISPLAY_NAME)));
+                newRecord.setFilePath(cursor.getString(cursor.getColumnIndex(InstanceProviderAPI.InstanceColumns.INSTANCE_FILE_PATH)));
+                newRecord.setCanEditWhenComplete(cursor.getString(cursor.getColumnIndex(InstanceProviderAPI.InstanceColumns.CAN_EDIT_WHEN_COMPLETE)));
+                cursor.close();
+            }
+            newRecords.add(newRecord);
+        }
+
+        // Drop old Table and create it again with new definition
+        db.execSQL("DROP TABLE IF EXISTS " + FormRecord.STORAGE_KEY);
+        TableBuilder builder = new TableBuilder(FormRecord.class);
+        db.execSQL(builder.getTableCreateString());
+
+        // Write to the new table
+        SqlStorage<FormRecord> newStorage = getFormRecordStorage(c, db, FormRecord.class);
+        for (FormRecord newRecord : newRecords) {
+            newStorage.write(newRecord);
+        }
+
+        // delete all instance provider entries
+        c.getContentResolver().delete(InstanceProviderAPI.InstanceColumns.CONTENT_URI, null, null);
```

Comment:
Same note as in the other migration that I think it's a better practice for us to track "success" for the transaction of this operation **before** this delete and not after, since it inherently can't be a part of the transaction